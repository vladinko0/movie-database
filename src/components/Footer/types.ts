export type FooterProps = {
  isLoading: boolean;
  pageNumber: number;
};
