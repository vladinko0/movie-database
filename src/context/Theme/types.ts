export interface Theme {
  cornflowerBlue: string;
  frenchPass: string;
  shineGray: string;
  white: string;
}

export type ThemeContextType = {
  theme: Theme;
};

export interface ThemeProps {
  children: React.ReactNode;
}
