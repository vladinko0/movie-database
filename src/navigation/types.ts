import {SCREENS} from './constants';
import {NavigationProp} from '@react-navigation/native';
import {TMovieItem} from '../screens/search/types';

export type RootStackParamList = {
  [SCREENS.HOME]: undefined;
  [SCREENS.SEARCH]: undefined;
  [SCREENS.FAVOURITES]: undefined;
  [SCREENS.DETAIL]: {movie: TMovieItem};
};

export type StackNavigation = NavigationProp<RootStackParamList>;
