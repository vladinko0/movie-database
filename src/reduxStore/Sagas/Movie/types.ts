import {
  DetailRequestBody,
  MovieDetailData,
  DetailFailureCallback,
  SearchRequestBody,
  SearchData,
  SearchFailureCallback,
} from '../../slices/types';

export type MovieDetailSaga = {
  requestBody: DetailRequestBody;
  successCallback: (data: MovieDetailData) => void;
  failureCallback: DetailFailureCallback;
};

export type SearchMoviesSaga = {
  requestBody: SearchRequestBody;
  successCallback: (data: SearchData) => void;
  failureCallback: SearchFailureCallback;
};
