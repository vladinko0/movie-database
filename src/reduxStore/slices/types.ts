import {TMovieItem} from '../../screens/search/types';

export interface MovieSliceProps {
  favourites: TMovieItem[];
}

export type FavouriteMovieAction = {
  payload: TMovieItem[];
};

export type FavouriteMovieState = {
  favourites: TMovieItem[];
};

export type SearchRequestBody = {
  searchString: string;
  page: number;
};
export type DetailRequestBody = {
  i: string;
};

export type MovieDetailData = {
  Title: string;
  Poster: string;
  Year: string;
  Genre: string;
  Plot: string;
};
export type MovieDetailResponse = {
  data: MovieDetailData | null;
  status: number;
  error?: string;
};

export type SearchData = {
  Search: TMovieItem[];
  totalResults: string;
};
export type SearchMovieResponse = {
  data: SearchData | null;
  status: number;
  error?: string;
};

export type DetailSuccessCallback = (response: MovieDetailResponse) => void;
export type DetailFailureCallback = (error: any) => void;
export type SearchSuccessCallback = (response: SearchMovieResponse) => void;
export type SearchFailureCallback = DetailFailureCallback;
