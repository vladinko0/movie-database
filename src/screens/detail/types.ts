import type {RouteProp} from '@react-navigation/native';
import {RootStackParamList} from '../../navigation/types';
import {SCREENS} from '../../navigation/constants';

export type DetailProps = {
  route: RouteProp<RootStackParamList, SCREENS.DETAIL>;
};

export type TDetail = {
  title?: string;
  poster?: string;
  year?: string;
  genre?: string;
  plot?: string;
};
