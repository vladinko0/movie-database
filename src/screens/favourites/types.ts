import {FooterProps} from '../../components/Footer/types';
import {RenderMovieItem} from '../search/types';

export type FavouritesProps = {
  renderFooter?: (props: FooterProps) => JSX.Element;
  renderMovieItem?: (props: RenderMovieItem) => JSX.Element;
};
