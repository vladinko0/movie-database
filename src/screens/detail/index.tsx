import React, {FC, memo, useEffect, useState, useRef, useCallback} from 'react';
import {View, ScrollView, Alert, Text, TouchableOpacity} from 'react-native';
import {useTheme} from '../../context/Theme';
import {styles} from './styles';
import LinearGradient from 'react-native-linear-gradient';
import {DetailProps, TDetail} from './types';
import {MovieDetailResponse} from '../../reduxStore/slices/types';
import {
  requestMovieDetails,
  setFavouriteMovie,
  useAppSelector,
  useAppDispatch,
} from '../../reduxStore';
import FastImage, {Source} from 'react-native-fast-image';
import Loader from '../../components/Loader';
import {fullStar, emptyStar} from '../../assets';

const Detail: FC<DetailProps> = ({route}) => {
  const movie = route.params.movie;
  const {theme} = useTheme();
  const dispatch = useAppDispatch();
  const [detail, setDetail] = useState<TDetail>({
    title: '',
    poster: '',
    year: '',
    genre: '',
    plot: '',
  });
  const [starSource, setStarSource] = useState<Source>(emptyStar);
  const isLoading = useRef<boolean>(true);
  const favourites = useAppSelector(state => state?.movieSlice.favourites);

  const getDetails = useCallback((): void => {
    const requestBody = {
      i: movie.imdbID,
    };
    dispatch(
      requestMovieDetails(
        requestBody,
        (response: MovieDetailResponse) => {
          isLoading.current = false;
          setDetail({
            title: response.data?.Title,
            poster: response.data?.Poster,
            year: response.data?.Year,
            genre: response.data?.Genre,
            plot: response.data?.Plot,
          });
        },
        error => {
          console.warn('Movie detail Error: ', error);
          Alert.alert(error);
        },
      ),
    );
  }, [dispatch, movie.imdbID]);

  const checkFavourites = useCallback((): void => {
    if (favourites.length && favourites.some(e => e.imdbID === movie.imdbID)) {
      setStarSource(fullStar);
    }
  }, [favourites, movie.imdbID]);

  useEffect(() => {
    getDetails();
    checkFavourites();
  }, [getDetails, checkFavourites]);

  const onStarPress = useCallback((): void => {
    if (starSource === emptyStar) {
      setStarSource(fullStar);
      if (
        !favourites.length ||
        !favourites.some(e => e.imdbID === movie.imdbID)
      ) {
        dispatch(setFavouriteMovie([...favourites, movie]));
      }
    } else {
      const newFavourites = favourites.filter(
        item => item.imdbID !== movie.imdbID,
      );
      dispatch(setFavouriteMovie(newFavourites));
      setStarSource(emptyStar);
    }
  }, [starSource, favourites, dispatch, movie]);

  return (
    <ScrollView style={styles.mainContainer}>
      <LinearGradient
        colors={[theme.cornflowerBlue, theme.frenchPass]}
        style={styles.gradient}>
        {isLoading.current ? (
          <Loader />
        ) : (
          <>
            <FastImage
              style={styles.image}
              source={{
                uri: detail.poster,
              }}
            />
            <View style={styles.textsView}>
              <View style={styles.titleView}>
                <Text style={styles.titleText}>{detail.title}</Text>
                <TouchableOpacity onPress={onStarPress}>
                  <FastImage style={styles.star} source={starSource} />
                </TouchableOpacity>
              </View>
              <Text style={styles.genreText}>
                {detail.year} / {detail.genre}
              </Text>
              <Text style={styles.genreText}>Plot:</Text>
              <Text style={styles.genreText}>{detail.plot}</Text>
            </View>
          </>
        )}
      </LinearGradient>
    </ScrollView>
  );
};

export default memo(Detail);
