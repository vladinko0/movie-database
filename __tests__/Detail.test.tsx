import 'react-native';
import React from 'react';
import Detail from '../src/screens/detail';
import renderer from 'react-test-renderer';
import {DetailProps} from '../src/screens/detail/types';

jest.mock('@react-native-async-storage/async-storage', () => ({
  AsyncStorage: () => {},
}));
jest.mock('redux-persist', () => ({
  persistStore: () => {},
  persistReducer: () => {},
}));
jest.mock('@reduxjs/toolkit', () => ({
  configureStore: () => {},
  createSlice: () => {},
}));
jest.mock('../src/reduxStore', () => ({
  useAppDispatch: () => {},
  useAppSelector: () => [],
}));

const mockRoute = {
  params: {
    movie: {
      Poster:
        'https://m.media-amazon.com/images/M/MV5BNmU0NmExYWYtY2E1YS00YWFlLTg5YmYtMjE2YWJkODRjNjk2XkEyXkFqcGdeQXVyNTU1MTE2MTI@._V1_SX300.jpg',
      Title: 'AAA Australian Open of Surfing',
      Type: 'movie',
      Year: '2016',
      imdbID: 'tt15176324',
    },
  },
} as DetailProps['route'];

it('renders correctly Details screen', () => {
  const detail = renderer.create(<Detail route={mockRoute} />);
  expect(detail).toMatchSnapshot();
});
