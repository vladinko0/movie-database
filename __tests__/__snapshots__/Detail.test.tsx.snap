// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Details screen 1`] = `
<RCTScrollView
  style={
    {
      "flex": 1,
    }
  }
>
  <View>
    <BVLinearGradient
      colors={
        [
          4283866105,
          4289848060,
        ]
      }
      endPoint={
        {
          "x": 0.5,
          "y": 1,
        }
      }
      locations={null}
      startPoint={
        {
          "x": 0.5,
          "y": 0,
        }
      }
      style={
        {
          "flex": 1,
          "minHeight": 800,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <ActivityIndicator />
      </View>
    </BVLinearGradient>
  </View>
</RCTScrollView>
`;
